### PR TITLE
fix: Tier module fixes and improvements (part 4)

### PR DESCRIPTION
**Pull request by iverc**
## Description
This PR aims to add a number of fixes and additional improvements on top of work done in https://github.com/sourcenetwork/sourcehub/pull/10

## Tasks
- [x] Add `UnlockingLockup`, refactor `Lockup`, `GenesisState`, and query messages.
- [x] Rename `unbondTime` to `completionTime` to match native staking module.
- [x] Add `creation_height` to  `MsgUnlockResponse`.
- [x] Store lockups / unlocking lockups separately in `GenesisState`.
- [x] Fix timestamp and calculation issues in `calculateProratedCredit`.
- [x] Improve logic related to storing/retrieving lockups and unlocking lockups. 
- [x] Handle `completionTime`/`unlockTime` edge case in `CompleteUnlocking`.
- [x] Refactor `proratedCredit`, add `TestCalculateProratedCredit` and `TestProratedCredit`.
- [x] Add multiple test suites, update/fix a bunch of tests.

### CODE REVIEW

**Review by Lodek — APPROVED** [review]
LGTM, I don't fully understand the credit calculations atm but overall the code looks good and the changes make sense after your code walkthrough. Good job Ivan, will review the remaining PRs tomorrow.

File: x/tier/keeper/calculate_credit.go [hunk 1]
```diff
@@ -41,3 +43,29 @@ func calculateCredit(rateList []types.Rate, lockedAmt, lockingAmt math.Int) math
...
+
+	if epochDurationMs == 0 {
+		return math.ZeroInt()
+	}
+
+	// This check is required because is possible that sinceCurrentEpoch can be greater than epochDuration
+	// (e.g. chain paused for longer than the epoch duration or misconfigured epoch duration).
+	if sinceCurrentEpoch < epochDurationMs {
+		credit = credit.MulRaw(sinceCurrentEpoch).QuoRaw(epochDurationMs)
+	}
```
Comment (jsimnz) [inline x/tier/keeper/calculate_credit.go:68]: praise: nice catch

File: x/tier/keeper/lockup.go
Comment (jsimnz) [inline x/tier/keeper/lockup.go:1]: question/future refactor.

We might want to be more aware of what functions are public/private on the Keeper, or more accurately what functions *need* to be public.

Reading through some of these Add/Subtract methods makes me feel like they are mostly necessary for the internal functionality to the keeper. Compared to the possibility of another dev using these unnecessarily or accidentally from a different package. 

Just a thought. Dont need to be updated in this PR

File: x/tier/types/keys.go [hunk 2]
```diff
@@ -89,8 +89,8 @@ func LockupKeyToAddresses(key []byte) (sdk.AccAddress, sdk.ValAddress) {
 	return delAddr, valAddr
 }
 
-// LockupKeyToAddressesAtHeight retreives delAddr, valAddr, and creationHeight from provided unlocking Lockup key.
-func LockupKeyToAddressesAtHeight(key []byte) (sdk.AccAddress, sdk.ValAddress, int64) {
```
Comment (jsimnz) [inline x/tier/types/keys.go:93]: suggestion: Are the `LockupKeyToAddressesAtHeight` not still handy util functions to have around?

**Review by jsimnz — APPROVED** [review]
***Really*** great work here. Like the proper separation between Lockup and UnlockingLockup state (albeit the names are a little confusing, understandably).

Have a few questions, but nothing that is a blocker, and one suggestion that is  up to you.

File: x/tier/keeper/lockup.go
Comment (iverc) [reply x/tier/keeper/lockup.go:1]: Agree, we should be more accurate on what needs to be public. Will make these changes in the upcoming PR. Thank you
